### PR TITLE
Update Proc

### PR DESCRIPTION
**Pull request by sampersand**
This PR updates `Proc`.

### CODE REVIEW

**Comment by sampersand** [comment]
Not sure what [this error](https://github.com/ruby/rbs/actions/runs/10024287064/job/27706196641?pr=1937) is for...

File: core/proc.rbs [hunk 1]
```diff
@@ -288,8 +288,184 @@
 #
 # Numbered parameters were introduced in Ruby 2.7.
 #
-class Proc < Object
-  def clone: () -> self
+class Proc
+  interface _Callable
+    def call: (*untyped, **untyped) ?{ (*untyped, **untyped) -> untyped } -> untyped
+  end
```
Comment (ParadoxV5) [inline core/proc.rbs:294]:
```suggestion
  interface _Call
    def call: (?) -> untyped
  end
```

File: core/proc.rbs [hunk 2]
```diff
@@ -288,8 +288,184 @@
...
+  # Creates a new Proc object, bound to the current context.
+  #
+  #     proc = Proc.new { "hello" }
+  #     proc.call   #=> "hello"
+  #
+  # Raises ArgumentError if called without a block.
+  #
+  #     Proc.new    #=> ArgumentError
+  #
+  def self.new: () { (*untyped, **untyped) -> untyped } -> instance
```
Comment (ParadoxV5) [inline core/proc.rbs:309]:
```suggestion
  def self.new: () { (?) -> untyped } -> instance
```

File: core/proc.rbs [hunk 3]
```diff
@@ -379,7 +555,7 @@ class Proc < Object
   #
   # See also Proc#lambda?.
   #
-  def call: (*untyped arg0) -> untyped
+  def call: (*untyped, **untyped) ?{ (*untyped, **untyped) -> untyped } -> untyped
```
Comment (ParadoxV5) [inline core/proc.rbs:558]: This is exactly what your #1622 was asking for.
```suggestion
  def call: (?) -> untyped
```

**Review by ParadoxV5 — COMMENTED** [review]
We have #1806 now.

File: core/proc.rbs (hunk 1 above)
Comment (sampersand) [reply core/proc.rbs:294]: Oh my, is this a new development I don't know about?

File: core/proc.rbs [hunk 4]
```diff
@@ -288,8 +288,184 @@
 #
 # Numbered parameters were introduced in Ruby 2.7.
 #
-class Proc < Object
-  def clone: () -> self
+class Proc
+  interface _Callable
+    def call: (?) -> untyped
+  end
```
Comment (ParadoxV5) [inline core/proc.rbs:294]: * Resolves https://github.com/ruby/rbs/issues/420#issue-723381303
  * but not https://github.com/ruby/rbs/issues/420#issuecomment-739523110 (https://github.com/ruby/rbs/discussions/1639#discussioncomment-7646487)

**Review by soutaro — APPROVED** [review]
Thanks! 🎉

**Comment by soutaro** [comment]
Rebased on the top of #1335, and added a commits for runtime type checker.

**Pull request merged**